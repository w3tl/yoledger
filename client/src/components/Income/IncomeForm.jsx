import React from 'react';
import PropTypes from 'prop-types';
import { TextInput, Button } from '../Elements';
import { incomePropType } from './propTypes';

class IncomeForm extends React.Component {
  constructor(props) {
    super(props);
    const { income } = props;
    this.state = {
      isCreate: !income.id,
      name: income.name,
    };
  }

  handleChange = name => ({ target }) => {
    this.setState({
      [name]: target.value,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { name } = this.state;
    const { onSave } = this.props;
    onSave({
      name,
    });
  }

  render() {
    const { isCreate, name } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        {isCreate && <TextInput label="Name" value={name} onChange={this.handleChange('name')} />}
        {!isCreate && <h4>{name}</h4>}
        {isCreate && (
          <Button disabled={name.length === 0} type="submit">
            Save
          </Button>)}
      </form>
    );
  }
}

IncomeForm.propTypes = {
  income: incomePropType,
  onSave: PropTypes.func,
};
IncomeForm.defaultProps = {
  income: {
    id: null,
    name: '',
  },
  onSave: () => {},
};

export default IncomeForm;
