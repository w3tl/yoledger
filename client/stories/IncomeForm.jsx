import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import IncomeForm from '../src/components/Income/IncomeForm';
import incomes from '../src/components/Income/mockData';

storiesOf('Income/Form', module)
  .add('create', () => <IncomeForm onSave={action('onSave')} />)
  .add('edit', () => <IncomeForm income={incomes[0]} onSave={action('onSave')} />);
