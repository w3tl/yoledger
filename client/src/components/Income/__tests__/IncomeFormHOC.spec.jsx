/* eslint-disable react/prop-types */
import React from 'react';
import { mount } from 'enzyme';
import formWithData from '../IncomeFormHOC';
import IncomeForm from '../IncomeForm';
import {
  wait, withProvider, testLoadingState,
} from '../../testHelpers/index';

jest.mock('react-router-dom', () => ({ // eslint-disable-next-line react/prop-types
  Redirect: ({ to }) => <a href={to}>Redirect</a>,
}));

const ComponentWithData = formWithData(IncomeForm);
const ComponentWithoutLocationState = withProvider(() => (
  <ComponentWithData />
));
const ComponentWithLocationState = withProvider(() => (
  <ComponentWithData location={{ state: { id: '31' } }} />
));

describe('IncomeFormHOC', () => {
  describe('with router id params', () => {
    testLoadingState(<ComponentWithLocationState />);

    it('should contain props', async () => {
      const wrapper = mount(<ComponentWithLocationState />);
      await wait();
      wrapper.update();
      const form = wrapper.find('IncomeForm');
      expect(form.prop('income').name).toBeDefined();
      expect(form.prop('onSave')).toBeDefined();
    });
  });

  describe('without router id params', () => {
    // NOTE: skip query without id in location
    // testLoadingState(<ComponentWithLocationState />);

    it('should successfully add expense', async () => {
      const wrapper = mount(<ComponentWithoutLocationState />);
      await wait();
      wrapper.update();
      wrapper.find('input').simulate('change', { target: { value: 'new account' } });
      wrapper.find('form').simulate('submit');
      await wait(1);
      wrapper.update();
      expect(wrapper.find('Redirect').prop('to')).toMatchSnapshot();
    });
  });
});
