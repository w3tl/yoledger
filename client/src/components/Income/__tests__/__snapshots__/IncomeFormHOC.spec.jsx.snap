// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomeFormHOC without router id params should successfully add expense 1`] = `
Object {
  "pathname": "/incomes/view",
  "state": Object {
    "id": "666",
  },
}
`;
